src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/W3CPropagator.cs

```diff
@@ -0,0 +1,745 @@
+                        vsb.Append(Replacement);
+                        i += 2;
+                        continue;
+                    }
+
+                    int cp = (((b0 & 0x07) << 18) | ((b1 & 0x3F) << 12) | ((b2 & 0x3F) << 6) | ((b3 & 0x3F)));
+
+                    if (cp < 0x10000 || cp > 0x10FFFF)
+                    {
+                        // Malformed utf-8 sequence. emit U+FFFD and continue
+                        vsb.Append(Replacement);
+                        i += 2;
+                        continue;
+                    }
+
+                    cp -= 0x10000;
+                    vsb.Append((char)((cp >> 10) + 0xD800));
+                    vsb.Append((char)((cp & 0x3FF) + 0xDC00));
+
+                    i += 11;
+                    continue;
+                }
+
+                // invalid byte sequence. emit U+FFFD and continue
+                vsb.Append(Replacement);
+                i += 2;
+            }
+
+            value = vsb.ToString();
+            return true;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static bool TryDecodeEscapedByte(ReadOnlySpan<char> span, out byte value)
+        {
+            Debug.Assert(span.Length > 0 && span[0] == Percent);
+
+            if (span.Length < 3 || !TryDecodeHexDigit(span[1], out byte byte1) || !TryDecodeHexDigit(span[2], out byte byte2))
+            {
+                value = 0;
+                return false;
+            }
+
+            value = (byte)(((uint)byte1 << 4) + byte2);
+            return true;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static bool TryDecodeHexDigit(char c, out byte value)
+        {
+            if ((uint)(c - '0') <= ('9' - '0'))
+            {
+                value = (byte)(c - '0');
+                return true;
+            }
+
+            if ((uint)(c- 'a') <= ('f' - 'a'))
+            {
+                value = (byte)((c - 'a') + 10);
+                return true;
+            }
+
+            if ((uint)(c - 'A') <= ('F' - 'A'))
+            {
+                value = (byte)((c - 'A') + 10);
+                return true;
+            }
+
+            value = 0;
+            return false;
+        }
+
+        // Allowed baggage key characters:
+        //  tchar = "!" / "#" / "$" / "%" / "&" / "'" / "*" / "+" / "-" / "." / "^" / "_" / "`" / "|" / "~" / DIGIT / ALPHA
+        //  DIGIT = 0-9
+        //  ALPHA = A-Z / a-z
+        private static ulong[] InvalidBaggageKeyCharsMask = [0xFC009305FFFFFFFF, 0x2800000038000001];
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static bool IsInvalidBaggageKeyCharacter(char c)
+        {
+            if (c >= 0x07F) // key support only ascii characters according to the W3C specs
+            {
+                return true;
+            }
+
+            return (InvalidBaggageKeyCharsMask[c >> 6] & (ulong)((ulong)1 << ((int)c & 63))) != 0;
+        }
```

Comment:
Can we replace all of these character checks with `SearchValues`?

The logic is written in a less readable way for performance reasons, but we're leaving a 10x difference on the table. You can avoid hardcoding all the bitmaps by doing something like
```c#
private const string ValidBaggageChars = "...";
#if NET
private static readonly SearchValues<char> s_validBaggageChars = SearchValues.Create(ValidBaggageChars);
private static bool IsInvalidBaggageKey(ReadOnlySpan<char> span) => span.ContainsAnyExcept(s_validBaggageChars);
#else
private static bool IsInvalidBaggageKey(ReadOnlySpan<char> span) => span.ContainsAnyExcept(ValidBaggageChars);
#endif
```

With a trivial polyfill for framework.

(same for all of these checks)

Reply:
I have applied using `SearchValues` in .NET but kept using the mask values in the poly fill for performance's sake. please let me know if you think otherwise.

Reply:
That seems fine.
In the future we should ideally look at polyfilling `SearchValues` itself to reduce the duplication.